### DISCUSSION TITLE
Signalr client frequently getting 404 when trying to connect to ALB in AWS

### DISCUSSION THREAD

**Discussion by shaykeren** [dotnet/aspnetcore#23350] [General]
Trying to connect Signalr client written with c# (netcoreapp 3.1 console application) toSignalr service running on two instances of WebAPI (netcoreapp3.1) behind application load balancer in AWS.
Frequently I'm getting the following error:

_Unable to connect to the server with any of the available transports. (WebSockets failed: The server returned status code '404' when status code '101' was expected.) (ServerSentEvents failed: Response status code does not indicate success: 404 (Not Found).) (LongPolling failed: Response status code does not indicate success: 404 (Not Found).)_

### I noticed the following behaviors:
1. Forcing my client to use WebSockets and with SkipNegotiation=true **[Works with no errors!]**
2. Forcing my client to use LongPolling or ServerSentEvents  and configure my application load balancer with Stickiness=true **[Works with not errors!]**
3. Configure my application load balancer with Stickiness=true  and using the following code(without specifying any transport)
`var connection = new HubConnectionBuilder()
    .WithUrl($"{host}/hub/agent")
    .WithAutomaticReconnect()
    .Build();
     await connection.StartAsync();` **[Works with no error!]** **BUT** the selected transport will be set to **ServerSentEvents** all the time.
My Question is Why Can't I use WebSocket without defining the SkipNegotiation=true in my environment?

Thanks

**Comment by davidfowl** [comment]
That's strange, does it fallback from websockets?

**Reply by shaykeren** [reply]
The third section - Yes, the selected transport type will be set to SSE.
What am I doing wrong?

**Reply by shaykeren** [reply]
Why Can't I use Websocket with Negotiation when using  Signalr behind LoadBalancer?

**Reply by davidfowl** [reply]
It should work fine, I'm not sure off the top of my head why it's not working. The only thing I could think of is that in that configuration somehow websockets aren't enabled. Can you turn on the client side logs https://docs.microsoft.com/en-us/aspnet/core/signalr/diagnostics?view=aspnetcore-3.1#javascript-client-logging?

**Reply by BrennanConroy** [reply]
@shaykeren Would it be possible for you to capture the headers/cookies of the requests and responses to/from the client? It would be nice to see the sticky session value for the /negotiate and the WebSocket request to see if they exist and match.

I'm thinking that maybe ALB ignores the sticky session header on WebSockets requests because WebSockets are "always sticky" and so the WebSocket might be redirected to any server.

**Reply by davidfowl** [reply]
@shaykeren what client are you using?

**Reply by davidfowl** [reply]
Ah just answered my own question!

**Reply by shaykeren** [reply]
Dotnet client (core)

**Comment by normj** [comment]
What compute power are you putting behind the load balancer? Containers, Windows EC2 instances or Linux EC2 instances? If it is a Linux environment is the load balancer going straight to Kestrel or do you have an Nginx reverse proxy in between?

**Reply by shaykeren** [reply]
Alb straight to kestrel (2 ec2 instances behind the alb)

**Comment by normj** [comment]
I'm able to reproduce the problem and it does seem to be error with stick sessions not being honored correctly since I can't reproduce the issue when their is only one instance behind the load balancer. 

It looks to me like the .NET SignalR client isn't carrying forward the sticky sessions cookie on the upgrade to websocket request.

Here is the raw request for the POST negotiate:
```
POST http://signalrtest-linux-alb.us-west-2.elasticbeanstalk.com/ChatHub/negotiate?negotiateVersion=1 HTTP/1.1
Host: signalrtest-linux-alb.us-west-2.elasticbeanstalk.com
User-Agent: Microsoft.AspNetCore.Http.Connections.Client/3.1.5
X-Requested-With: XMLHttpRequest
Content-Length: 0
```
and the response contains cookies ALB uses for tracking sticky sessions
```
HTTP/1.1 200 OK
Date: Sat, 27 Jun 2020 00:39:53 GMT
Content-Type: application/json
Content-Length: 316
Connection: keep-alive
Set-Cookie: AWSALB=jM6lmN2pMQmLSACC6CKsvXSCpZg0j6R/kSG/VldkRNXvp96qrXS75qJXgYCxdQ9CBKnW5eiCiHtzEN6RRhBTRQRc3PKWt2lTxxsb/L2+hMU3n75BpwcNTG2T9l9Q; Expires=Sat, 04 Jul 2020 00:39:53 GMT; Path=/
Set-Cookie: AWSALBCORS=jM6lmN2pMQmLSACC6CKsvXSCpZg0j6R/kSG/VldkRNXvp96qrXS75qJXgYCxdQ9CBKnW5eiCiHtzEN6RRhBTRQRc3PKWt2lTxxsb/L2+hMU3n75BpwcNTG2T9l9Q; Expires=Sat, 04 Jul 2020 00:39:53 GMT; Path=/; SameSite=None
Server: nginx/1.16.1

{"negotiateVersion":1,"connectionId":"Qyt5WO9f2uKS6VCwBEaS_w","connectionToken":"NDYZsRae-6sflSEgNlB0xQ","availableTransports":[{"transport":"WebSockets","transferFormats":["Text","Binary"]},{"transport":"ServerSentEvents","transferFormats":["Text"]},{"transport":"LongPolling","transferFormats":["Text","Binary"]}]}
```

The response contains the `Set-Cookie` header for `AWSALB` which needs to be put on future request to maintain the sticky session but the following request to upgrade the connection does not contain the Cookie.

```
GET http://signalrtest-linux-alb.us-west-2.elasticbeanstalk.com/ChatHub?id=NDYZsRae-6sflSEgNlB0xQ HTTP/1.1
Host: signalrtest-linux-alb.us-west-2.elasticbeanstalk.com
X-Requested-With: XMLHttpRequest
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Version: 13
Sec-WebSocket-Key: 26Na1dMNJUO7Q+tNq62N+g==
```

Here is the failure response
```
HTTP/1.1 404 Not Found
Date: Sat, 27 Jun 2020 00:39:54 GMT
Transfer-Encoding: chunked
Connection: keep-alive
Set-Cookie: AWSALB=ajki7DwusIlIWnCRp5gNCpcDrlmjzIPvl9UwxZFy9qCnrqlh158PhjV9zc+OS10Js1bXuCOOmzCmYutzfJIbR76OhU/0xn9KtWNf6wSo2tzu1rY1T19gasxpyqPN; Expires=Sat, 04 Jul 2020 00:39:54 GMT; Path=/
Set-Cookie: AWSALBCORS=ajki7DwusIlIWnCRp5gNCpcDrlmjzIPvl9UwxZFy9qCnrqlh158PhjV9zc+OS10Js1bXuCOOmzCmYutzfJIbR76OhU/0xn9KtWNf6wSo2tzu1rY1T19gasxpyqPN; Expires=Sat, 04 Jul 2020 00:39:54 GMT; Path=/; SameSite=None
Server: nginx/1.16.1

1a
No Connection with that ID
0
```

I do see subsequent request passing along the Cookies but I assume that is when it is falling back to other protocol.

```
POST http://signalrtest-linux-alb.us-west-2.elasticbeanstalk.com/ChatHub/negotiate?negotiateVersion=1 HTTP/1.1
Host: signalrtest-linux-alb.us-west-2.elasticbeanstalk.com
User-Agent: Microsoft.AspNetCore.Http.Connections.Client/3.1.5
X-Requested-With: XMLHttpRequest
Cookie: AWSALB=jM6lmN2pMQmLSACC6CKsvXSCpZg0j6R/kSG/VldkRNXvp96qrXS75qJXgYCxdQ9CBKnW5eiCiHtzEN6RRhBTRQRc3PKWt2lTxxsb/L2+hMU3n75BpwcNTG2T9l9Q; AWSALBCORS=jM6lmN2pMQmLSACC6CKsvXSCpZg0j6R/kSG/VldkRNXvp96qrXS75qJXgYCxdQ9CBKnW5eiCiHtzEN6RRhBTRQRc3PKWt2lTxxsb/L2+hMU3n75BpwcNTG2T9l9Q
Content-Length: 0

```

**Comment by shaykeren** [comment]
So the stickiness is mandatory only for the Signalr  negotiation phase?
Is it a real bug of the dotnet core client?

**Reply by davidfowl** [reply]
Yes it looks like a bug... You might be able to work around it. We're looking into it.

**Reply by shaykeren** [reply]
Thank, waiting for workaround..

**Reply by BrennanConroy** [reply]
One possible workaround would be to make an HTTP request to the server and grab the cookies off the response and set them on the options when constructing the HubConnection.

```c#
var cookies = new CookieContainer();
var handler = new HttpClientHandler();
handler.CookieContainer = cookies;
var httpClient = new HttpClient(handler);
httpClient.BaseAddress = new Uri("http://siteurl");
_ = await httpClient.GetAsync("");

var hubConnection = new HubConnectionBuilder()
    .WithUrl(uri, o =>
    {
        foreach (var cookie in cookies.GetCookies(new Uri("http://siteurl")))
        {
            o.Cookies.Add((Cookie)cookie);
        }
    })
    .Build();
```

**Reply by shaykeren** [reply]
Is there a bug on this one? should I open one?

**Reply by BrennanConroy** [reply]
No bug, feel free to open one.

**Reply by davidfowl** [reply]
I just realized we could have turned this into an issue 😄
